sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Google.Protobuf.WellKnownTypes;
+using Pulumi.Serialization;
+using Pulumirpc;
+
+namespace Pulumi
+{
+    public class InvokeOptions
+    {
+        /// <summary>
+        /// An optional parent to use for default options for this invoke (e.g. the default provider
+        /// to use).
+        /// </summary>
+        public Resource? Parent { get; set; }
+
+        /// <summary>
+        /// An optional provider to use for this invocation. If no provider is supplied, the default
+        /// provider for the invoked function's package will be used.
+        /// </summary>
+        public ProviderResource? Provider { get; set; }
+        /// <summary>
+        /// An optional version, corresponding to the version of the provider plugin that should be
+        /// used when performing this invoke.
+        /// </summary>
+        public string? Version { get; set; }
+    }
+
+    public sealed partial class Deployment
+    {
+        public async Task<T> InvokeAsync<T>(

```

Comment:
Do we plan on only supporting `InvokeAsync` in the core? IIRC the user can synchronously wait on the result without issue, but I wonder if we should also have synchronous `Invoke`.

Reply:
>  but I wonder if we should also have synchronous Invoke.

We could have this.  Though we have to decide on the semantics.  Is `Invoke`:
1. Exactly the same as `InvokeAsync().Wait()`? or is it
2. An impl that does no blocking on async work, but then requires all input data to be synchronously known.

I didn't particularly want to answer that question now, so i left it as just InvokeAsync, with teh user able to just block on it if they want.

Reply:
Why would people need a sync one? Is there a guideline saying we should or shouldn't do this?
I would default to async all the way.

Reply:
Beware of introducing "sync over async" issues. Keep it simple and async all the way?

Reply:
Beware of introducing "sync over async" issues. Keep it simple and async all the way?

Reply:
> Why would people need a sync one? 

In general, because it's simpler, and because if they're currently synchronous, this doesn't force them to become async.

That said, with the dotnet model we're pushing, we're heavily pushing people to be async from teh start.  i.e. an `async-Main` method.  an async `Deployment.Run` method, etc.  so it should ideally be non-onerous to be async here.  

So i'm going to stick with async for now absent pressing need for the near future.